src/lib.rs

```diff
@@ -338,128 +579,218 @@ impl Graph {
-                    output_assignments
-                        .get(&(node, *port))
-                        .map(|(buffer, _)| (*port, *buffer))
-                })
-                .collect::<Vec<_>>();
-            Scheduled {
-                node,
-                inputs,
-                outputs,
-            }
-        })
+        // This won't panic because this is always `Some` on the user's end.
+        let mut mem_cache = self.mem_cache.take().unwrap();
+
+        let mut _scheduled = mem_cache.scheduled.take().unwrap_or_default();
+
+        mem_cache.all_latencies.clear();
+        mem_cache.all_latencies.resize(self.node_count(), None);
+        mem_cache.delays.clear();
+
+        mem_cache.allocator.clear();
+
+        mem_cache.input_assignments.clear();
+        mem_cache.output_assignments.clear();
+
+        let mut scheduled_nodes = mem_cache.scheduled_nodes.take().unwrap_or_default();
+        scheduled_nodes.clear();
+
+        // This is arguably quite expensive, but the reason is that we don't want to add any delay
+        // compensation nodes to the user's graph (We only want to add them to the schedule).
+        // The fact that we're memory caching the previous graph should help some.
+        let mut delay_comp_graph =
+            if let Some(mut delay_comp_graph) = mem_cache.delay_comp_graph.take() {
+                *delay_comp_graph = self.clone();
+                delay_comp_graph
+            } else {
+                Box::new(self.clone())
+            };
+
+        let mut latencies = mem_cache.latencies.take().unwrap_or_default();
+        let mut deps = mem_cache.deps.take().unwrap_or_default();
+        let mut edges = mem_cache.edges.take().unwrap_or_default();
+        let mut insertions = mem_cache.insertions.take().unwrap_or_default();
+        let mut queue = mem_cache.walk_queue.take().unwrap_or_default();
+        let mut visisted = mem_cache.walk_visited.take().unwrap_or_default();
+
+        delay_comp_graph.walk_mut(
+            root,
+            &mut mem_cache,
+            &mut queue,
+            &mut visisted,
+            |graph, node, mem_cache| {
+                // Maybe use the log crate for this to avoid polluting the user's output?
+                // println!("compiling {}", graph.node_name(node).unwrap());
+
+                solve_latency_requirements(
+                    graph,
+                    node,
+                    mem_cache,
+                    &mut latencies,
+                    &mut deps,
+                    &mut edges,
+                    &mut insertions,
+                );
+                for insertion in insertions.iter() {
+                    solve_buffer_requirements(graph, *insertion, mem_cache);
+                    scheduled_nodes.push(*insertion);
+                }
+                solve_buffer_requirements(graph, node, mem_cache);
+                scheduled_nodes.push(node);
+            },
+        );
+
+        _scheduled = scheduled_nodes
```

Comment:
We can remove the _ prefix since this variable is used

Reply:
Yeah, I suppose it is used. Rust analyzer thought it wasn't for some reason.

Reply:
Apparently rustc flags this as a warning as well. I think it's because we're overwriting the value with `collect()` instead of reading it. I'll change it so it's more clear.